packages/flutter/lib/src/cupertino/route.dart

```diff
@@ -453,44 +453,95 @@ class CupertinoPageTransition extends StatelessWidget {
-  ///  * `linearTransition` is whether to perform the secondary transition linearly.
+  final Animation<double> secondaryRouteAnimation;
+
+  ///  * `linearTransition` is whether to perform the transitions linearly.
   ///    Used to precisely track back gesture drags.
-  CupertinoFullscreenDialogTransition({
-    super.key,
-    required Animation<double> primaryRouteAnimation,
-    required Animation<double> secondaryRouteAnimation,
-    required this.child,
-    required bool linearTransition,
-  }) : _positionAnimation = CurvedAnimation(
-         parent: primaryRouteAnimation,
+  final bool linearTransition;
+
+  /// The widget below this widget in the tree.
+  final Widget child;
+
+  @override
+  State<CupertinoFullscreenDialogTransition> createState() => _CupertinoFullscreenDialogTransitionState();
+}
+
+class _CupertinoFullscreenDialogTransitionState extends State<CupertinoFullscreenDialogTransition> {
+
+    // When this page is coming in to cover another page.
+  late Animation<Offset> _primaryPositionAnimation;
+  // When this page is becoming covered by another page.
+  late Animation<Offset> _secondaryPositionAnimation;
+  // Curve of primary page which is coming in to cover another page.
+  CurvedAnimation? _primaryPositionCurve;
+  // Curve of secondary page which is becoming covered by another page.
+  CurvedAnimation? _secondaryPositionCurve;
+
+  @override
+  void initState() {
+    super.initState();
+    _setupAnimation();
+  }
+
+  @override
+  void didUpdateWidget(covariant CupertinoFullscreenDialogTransition oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (oldWidget.primaryRouteAnimation != widget.primaryRouteAnimation
+    || oldWidget.secondaryRouteAnimation != widget.secondaryRouteAnimation
+    || oldWidget.child != widget.child || oldWidget.linearTransition != widget.linearTransition) {
+    _disposeCurve();
+    _setupAnimation();
+    }
+  }
+
+   @override
+  void dispose() {
+    super.dispose();
+    _disposeCurve();
+  }
+
+  void _disposeCurve() {
+    _primaryPositionCurve?.dispose();
+    _secondaryPositionCurve?.dispose();
+  }
+
+  void _setupAnimation() {
+    _primaryPositionAnimation =  (_primaryPositionCurve = CurvedAnimation(
+         parent: widget.primaryRouteAnimation,
          curve: Curves.linearToEaseOut,
          // The curve must be flipped so that the reverse animation doesn't play
          // an ease-in curve, which iOS does not use.
          reverseCurve: Curves.linearToEaseOut.flipped,
-       ).drive(_kBottomUpTween),
+       )).drive(_kBottomUpTween);
        _secondaryPositionAnimation =
-           (linearTransition
-             ? secondaryRouteAnimation
-             : CurvedAnimation(
-                 parent: secondaryRouteAnimation,
+           (widget.linearTransition
+             ? widget.secondaryRouteAnimation
+             : _secondaryPositionCurve = CurvedAnimation(
+                 parent: widget.secondaryRouteAnimation,
```

Comment:
Here you are never reassigning `null` to the curved animation. This means you might dispose them twice